src/librustc_infer/traits/select.rs

```diff
@@ -1285,9 +1277,7 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
         result: &SelectionResult<'tcx, SelectionCandidate<'tcx>>,
     ) -> bool {
         match result {
-            Ok(Some(SelectionCandidate::ParamCandidate(trait_ref))) => {
-                !trait_ref.skip_binder().input_types().any(|t| t.walk().any(|t_| t_.is_ty_infer()))
-            }
+            Ok(Some(SelectionCandidate::ParamCandidate(trait_ref))) => !trait_ref.needs_infer(),
```

Comment:
This is a fix for https://github.com/rust-lang/rust/pull/55258#discussion_r383309878, I should make a separate commit but I didn't want to lose track of this.